# ChIP-seq scripts

Here reside scripts that bioinformaticians in Hoffmann lab and I wrote and contributed to analyze ChIP-seq data. These are pretty specific in terms of how to use them. The order of executing the scripts will be described below in detail. A goal of having all of the scripts here is to allow everyone in the lab to perform upstream of ChIP-seq analysis.

## Getting Started

Use 2trim.sh first. Then use algn2spp_test.sh. Use callpks.sh or callpks005.sh depending on FDR cutoff you want to use. Use makebw.sh to create sequencing depth normalized tracks. 

*Under development

### Directory Architecture
```
Under development
```

